corehq/apps/locations/permissions.py

```diff
@@ -239,27 +235,31 @@ def is_location_safe(view_fn, request, view_args, view_kwargs):
     request, view_args and kwargs are also needed because view_fn alone doesn't always
     contain enough information
     """
+    # Tastypie
+    if 'resource_name' in view_kwargs:
+        return view_kwargs['resource_name'] in LOCATION_SAFE_TASTYPIE_RESOURCES
+
+    # HQ report - get report class
     if getattr(view_fn, 'is_hq_report', False):
         dispatcher = view_fn.view_class
         report_class = dispatcher.get_report(view_kwargs['domain'], view_kwargs['report_slug'])
         view_fn = report_class
 
-    if getattr(view_fn, 'is_location_safe', False):
-        if isinstance(view_fn, type):
-            return _has_own_location_safe_attribute(view_fn)
-        if hasattr(view_fn, "view_class"):
-            return _has_own_location_safe_attribute(view_fn.view_class)
+    # Django view - get view class
+    if hasattr(view_fn, "view_class"):
+        view_fn = view_fn.view_class
+
+    if _get_own_property(view_fn, 'is_location_safe'):
         return True
-    if 'resource_name' in view_kwargs:
-        return view_kwargs['resource_name'] in LOCATION_SAFE_TASTYPIE_RESOURCES
-    if getattr(view_fn, '_conditionally_location_safe_function', False):
-        return view_fn._conditionally_location_safe_function(view_fn, request, *view_args, **view_kwargs)
+    conditional_fn = _get_own_property(view_fn, '_conditionally_location_safe_function')
+    if conditional_fn:
+        return conditional_fn(view_fn, request, *view_args, **view_kwargs)
```

Comment:
Does it matter that `view_fn` may have been reassigned to a class (`view_fn = view_fn.view_class`) above?

Reply:
That's by design - my goal here is to store `is_location_safe` or `_conditionally_location_safe_function` directly on the thing that handles the request, whether a function view, class view, or HQ report.  That's whatever we developers would consider the main model for handling a request.  Then the first part of this function can normalize to that, and the boolean or conditional check can be evaluated in one place.

I think `view_fn` is probably no longer the best name for that though - got any ideas?

Here's refactor of the above if it makes the meaning clearer:

```python
def _view_obj_is_safe(obj, request, *view_args, **view_kwargs):
    if _get_own_property(obj, 'is_location_safe'):
        return True
    conditional_fn = _get_own_property(obj, '_conditionally_location_safe_function')
    if conditional_fn:
        return conditional_fn(obj, request, *view_args, **view_kwargs)


def is_location_safe(view_fn, request, view_args, view_kwargs):
    if 'resource_name' in view_kwargs:  # Tastypie
        return view_kwargs['resource_name'] in LOCATION_SAFE_TASTYPIE_RESOURCES

    if getattr(view_fn, 'is_hq_report', False):  # HQ report
        dispatcher = view_fn.view_class
        report_class = dispatcher.get_report(view_kwargs['domain'], view_kwargs['report_slug'])
        return _view_obj_is_safe(report_class, request, *view_args, **view_kwargs)

    if hasattr(view_fn, "view_class"):  # Django view
        return _view_obj_is_safe(view_fn.view_class, request, *view_args, **view_kwargs)

    return _view_obj_is_safe(view_fn, request, *view_args, **view_kwargs)
```

Reply:
I should also note that there's a tight coupling between this and the decorator `location_safe`.  The approach I'm settling on is to make the decorator simpler - where it just sets a property on the object, and have this handle the complexity of the possible different types of objects.  If you look higher in the diff, you can see that the decorator used to set the property differently depending on what was being decorated.

Reply:
:+1: sounds good